pyartcd/pyartcd/pipelines/promote.py

```diff
@@ -375,6 +401,215 @@ def _reraise_if_not_permitted(self, err: VerificationError, code: str, permits:
+            # always return 4.m.0.
+            url = 'https://amd64.ocp.releases.ci.openshift.org/api/v1/releasestream/4-stable/latest'
+            full_url = f"{url}?in=>4.{prevMinor}.0-0+<4.{prevMinor}.1"
+            # See if the previous minor has GA'd yet; e.g. https://amd64.ocp.releases.ci.openshift.org/releasestream/4-stable/release/4.8.0
+            async with aiohttp.ClientSession() as session:
+                async with session.get(full_url) as response:
+                    text = await response.json()
+                    prevGA = text['name']
+                async with session.get(f"{rcURL}/releasestream/{stableStream}/release/{prevGA}") as response:
+                    if response.status == 200:
+                        # If prevGA is known to the release controller, compute the changelog html
+                        async with session.get(f"{rcURL}/changelog?from={prevGA}&to={release_name}&format=html") as response:
+                            text = await response.text()
+                            with open(outputDest, 'w') as f:
+                                f.write(await response.text())
+                        # Also collect the output in markdown for SD to consume
+                        async with session.get(f"{rcURL}/changelog?from={prevGA}&to={release_name}&format=html") as response:
+                            text = await response.text()
+                            with open(outputDestMd, 'w') as f:
+                                f.write(text)
+                    else:
+                        with open(outputDest, 'w') as f:
+                            f.write(f"<html><body><p>Changelog information cannot be computed for this release. Changelog information will be populated for new releases once {prevGA} is officially released.</p></body></html>")
+                        with open(outputDestMd, 'w') as f:
+                            f.write(f"Changelog information cannot be computed for this release. Changelog information will be populated for new releases once {prevGA} is officially released.")
+        except Exception as e:
+            self._logger.error("Error generating changelog for release")
+            raise e
+
+    def extract_opm(self, client_mirror_dir, release_name, operator_registry, arch):
+        current_path = os.getcwd()
+        binaries = ['opm']
+        platforms = ['linux']
+        if arch == 'x86_64':  # For x86_64, we have binaries for macOS and Windows
+            binaries += ['darwin-amd64-opm', 'windows-amd64-opm']
+            platforms += ['mac', 'windows']
+        path_args = []
+        for binary in binaries:
+            path_args.append(f'--path=/usr/bin/registry/{binary}:{client_mirror_dir}')
+        extract_release_binary(operator_registry, path_args)
+        # Compress binaries into tar.gz files and calculate sha256 digests
+        os.chdir(client_mirror_dir)
```

Comment:
Avoid `os.chdir` if possible. This could lead to race condition if we run those functions with multithread or coroutine in the future.

Reply:
if we don't use chdir it will make symlink contents on mirror incorrect like https://redhat-internal.slack.com/archives/CJARLA942/p1683773060618819

Reply:
This is not true. symlinks have nothing to do with current directory or chdir call. What you need to do is setting the symlink link target to a relative path instead of full physical path.